slowmovie.py

```diff
@@ -10,205 +10,372 @@
 # ** Waveshare library   **
 # *************************
 
-import os, time, sys, random 
-from PIL import Image
-import ffmpeg
-import argparse
-
-# Ensure this is the correct import for your particular screen 
-from waveshare_epd import epd7in5_V2
-
-def generate_frame(in_filename, out_filename, time, width, height):    
-    (
-        ffmpeg
-        .input(in_filename, ss=time)
-        .filter('scale', width, height, force_original_aspect_ratio=1)
-        .filter('pad', width, height, -1, -1)
-        .output(out_filename, vframes=1)              
-        .overwrite_output()
-        .run(capture_stdout=True, capture_stderr=True)
-    )
-
-def check_mp4(value):
-    if not value.endswith('.mp4'):
-        raise argparse.ArgumentTypeError("%s should be an .mp4 file" % value)
-    return value
-
-# Ensure this is the correct path to your video folder 
-viddir = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'Videos/')
-logdir = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'logs/')
-
-
-parser = argparse.ArgumentParser(description='SlowMovie Settings')
-parser.add_argument('-r', '--random', action='store_true', 
-    help="Random mode: chooses a random frame every refresh")
-parser.add_argument('-f', '--file', type=check_mp4,
-    help="Add a filename to start playing a specific film. Otherwise will pick a random file, and will move to another film randomly afterwards.")
-parser.add_argument('-d', '--delay',  default=120, 
-    help="Delay between screen updates, in seconds")
-parser.add_argument('-i', '--inc',  default=4, 
-    help="Number of frames skipped between screen updates")
-parser.add_argument('-s', '--start',  
-    help="Start at a specific frame")
+import os, time, sys, random, ffmpeg, argparse, signal
+from PIL import Image, ImageEnhance
+
+# Ensure this is the correct import for your particular screen
+from waveshare_epd import epd7in5_V2 as epd_driver
+
+def clamp(n, smallest, largest):
+    return max(smallest, min(n, largest))
+
+# Handle when the program is killed and exit gracefully
+def signal_handler(signum, frame):
```

Comment:
I noticed [this commit](https://github.com/tkameroski/SlowMovie/commit/1efaec0c5b8e9b2ee60382960c11c27d3bff5622) from @tkameroski. What's the difference between these approaches? Why does one work with ctrl-C? Thoughts anyone?

Reply:
Comparing the PR to this commit the only difference is the removal of ```epd_driver.epdconfig.module_init()``` right? 

The approach in this PR is what I'd call the _correct_ one based on how the waveshare library is constructed. If you don't call init() before the exit() function you get RunTime error when hitting CTL-C. The epd driver calls ```SPI.close()``` when you use the sleep() function. Because of this the exit handler is calling close() on an already closed object. You have to init first, and then close it properly. It's bit odd since you want all the other clean-up operations that exist in the ```module_exit()``` function to happen but the SPI call will thrown an error if you don't open it first. 

I tested this again with the PR code by removing the init() call and you get a RunTime error. Putting it back in you get a clean exist.

Reply:
Using try works too
```
def exithandler(signum, frame):
    try:
        epd_driver.epdconfig.module_exit()
    finally:
        sys.exit()
```

Reply:
I think the way signal_handler is currently defined (from @robweber's PR) matches @tkameroski's approach. And, as opposed to @missionfloyd's try: finally: approach, it calls `module(exit)` (plus the accompanying clean-up operations) every time instead of only when SPI is open. If I understand all that correctly. So I think this can be resolved with no changes.

Reply:
My way does call module_exit() every time. It'll obviously fail if SPI isn't open, thus the try statement. Anyway, that commit message is incorrect; it already worked with ctrl-c.
```
signal.signal(signal.SIGTERM, exithandler) # 'kill' / service stop
signal.signal(signal.SIGINT, exithandler) # Ctrl-C
```
What it really did is make it work with waveshare's recent updates, which close SPI every time sleep() is called. Same thing https://github.com/missionfloyd/SlowMovie/commit/01ff5cf3c0ff294f9d2842be27d5c0249b99bddf does.

Reply:
@missionfloyd  I don't think anyone is arguing that your version doesn't fail gracefully, it does. The difference is that if the SPI.close() method fails and then throws an error you miss any other processing that was done in the [module_exit() ](https://github.com/qubist/SlowMovie/blob/master/e-paper/rpi/python/lib/waveshare_epd/epdconfig.py#L81)function. By using ```module_init()``` first the ```module_exit()``` function gracefully exits and the rest of the operations complete. 

I totally agree the waveshare code should be better about this and honestly it's on them to make these kind of checks so methods don't fail but they don't.

Reply:
It doesn't need to do the rest of it because it'll have already been done. That's the whole reason it fails.

Reply:
I don't see any other place the GPIO gets closed properly prior to this. The SPI call happens first and that's where the error will happen.

Reply:
After every frame, [epd.sleep()](https://github.com/qubist/SlowMovie/blob/e1e36358cc78cfa9bd40f4231cdf46c8b2337c11/slowmovie.py#L376) is called, which in turn calls [epdconfig.module_exit()](https://github.com/waveshare/e-Paper/blob/af9694c602fb24d776cd0f0b3e073a477294e3d8/RaspberryPi_JetsonNano/python/lib/waveshare_epd/epd7in5b_V2.py#L170-L178). If you stop it while it's sending a frame, SPI is still open, so this doesn't fail.

Reply:
You're linking to code in two different repos. That's probably why our thoughts on this differ. The waveshare code, as it exists here in qubist:master, doesn't call module_exit() when you use sleep. It calls [module_sleep() ](https://github.com/qubist/SlowMovie/blob/e1e36358cc78cfa9bd40f4231cdf46c8b2337c11/e-paper/rpi/python/lib/waveshare_epd/epd7in5_V2.py#L168)which only closes SPI and that's it. The current version of this program bundles the waveshare lib with the ```slowmovie.py``` file. Probably not best practice but we can't make decisions in that code based on code we're not bundling with the it. 

We should probably test the main waveshare stuff again now that it seems to have incorporated fixes and think about unbundling it from this repo.

Reply:
Yeah, the waveshare driver here is outdated. It should probably be removed.